api/v1alpha1/driver_types.go

```diff
@@ -0,0 +1,268 @@
+	Plugin    *corev1.ResourceRequirements `json:"plugin,omitempty"`
+}
+
+type PluginSpec struct {
+	// Embedded common pods spec
+	PodCommonSpec `json:"inline"`
+
+	// Driver's plugin daemonset update strategy, supported values are OnDelete and RollingUpdate.
+	// Default value is RollingUpdate with MaxAvailabile set to 1
+	UpdateStrategy *appsv1.DaemonSetUpdateStrategy `json:"updateStrategy,omitempty"`
+
+	// Resource requirements for plugin's containers
+	Resources PluginResourcesSpec `json:"resources,omitempty"`
+
+	Volumes []corev1.Volume `json:"volumes,omitempty"`
+
+	// kubelet directory path, if kubelet configured to use other than /var/lib/kubelet path.
+	KubeletDirPath string `json:"kubeletDirPath,omitempty"`
+
+	// Control the host mount of /etc/selinux for csi plugin pods. Defaults to false
+	EnableSeLinuxHostMount *bool `json:"EnableSeLinuxHostMount,omitempty"`
+
+	// To indicate the image pull policy to be applied to all the containers in the csi driver pods.
+	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy"`
+}
+
+type ProvisionerResourcesSpec struct {
+	Attacher      *corev1.ResourceRequirements `json:"attacher,omitempty"`
+	Snapshotter   *corev1.ResourceRequirements `json:"snapshotter,omitempty"`
+	Resizer       *corev1.ResourceRequirements `json:"resizer,omitempty"`
+	Provisioner   *corev1.ResourceRequirements `json:"provisioner,omitempty"`
+	OMapGenerator *corev1.ResourceRequirements `json:"omapGenerator,omitempty"`
+	Liveness      *corev1.ResourceRequirements `json:"liveness,omitempty"`
+	Plugin        *corev1.ResourceRequirements `json:"plugin,omitempty"`
+}
+
+type ProvisionerSpec struct {
+	// Embedded common pods spec
+	PodCommonSpec `json:"inline"`
+
+	// Set replicas for csi provisioner deployment. Defaults to 2
+	Replicas *int32 `json:"provisionerReplicas,omitempty"`
+
+	// Resource requirements for provisioner's containers
+	Resources ProvisionerResourcesSpec `json:"resources,omitempty"`
+}
+
+type LivenessSpec struct {
+	// Port to expose liveness metrics
+	MetricsPort int `json:"metricsPort,omitempty"`
+}
+
+type LeaderElectionSpec struct {
+	// Duration in seconds that non-leader candidates will wait to force acquire leadership.
+	// Default to 137 seconds.
+	LeaseDuration int `json:"leaseDuration,omitempty"`
+
+	// Deadline in seconds that the acting leader will retry refreshing leadership before giving up.
+	// Defaults to 107 seconds.
+	RenewDeadline int `json:"renewDeadline,omitempty"`
+
+	// Retry Period in seconds the LeaderElector clients should wait between tries of actions.
+	// Defaults to 26 seconds.
+	RetryPeriod int `json:"retryPeriod,omitempty"`
+}
+
+type CephFsClientType string
+
+const (
+	KernelCephFsClient CephFsClientType = "kernel"
+	FuseCephFsClient   CephFsClientType = "fuse"
+)
+
+// DriverSpec defines the desired state of Driver
+type DriverSpec struct {
+	// Logging configuration for driver's pods
+	Logging *LogSpec `json:"logging,omitempty"`
```

Comment:
should we call this as log or log in with the json tag? `log` looks good to me.

Reply:
Why `log` and not `logging`? What is wrong with the current name?